File: ClassicApp/TableViewController.m
//
//  TableViewController.m
//  TableViewController
//
//  Created by Leo Dion on 8/30/21.
//

#import "TableViewController.h"
#import "TableViewCell.h"
#import "Facade.h"
#import "Member.h"

@interface TableViewController ()
@property (nonatomic, retain) NSArray *members;
@property (nonatomic, retain) NSError *error;
@end

@implementation TableViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  UINib * nib = [UINib nibWithNibName:@"TableViewCell" bundle:nil];
  [self.tableView registerNib:nib forCellReuseIdentifier:@"reuseIdentifier"];
  
  __weak TableViewController *weakSelf = self;
  [[Facade sharedObject] getMembers:^(NSArray *members, NSError *error) {
    weakSelf.members = members;
    weakSelf.error = error;
    [weakSelf.tableView reloadData];
  }];
}

#pragma mark - Table view data source


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  
  return self.members.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  
  TableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"reuseIdentifier" forIndexPath:indexPath];
  
  Member * member = [self.members objectAtIndex:indexPath.row];
  
  if (member && cell) {
    cell.nameLabel.text = member.firstName;
    cell.nameLabel.textColor = member.uiColor;

    NSLog(@"%@", member.id);
    NSLog(@"%@", member.firstName);
    NSLog(@"%@", member.lastName);
    NSLog(@"%@", member.imageURL);
    NSLog(@"%@", member.email);
    NSLog(@"%@", member.summary);
    NSLog(@"%@", member.createdAt);
    NSLog(@"%lu", (unsigned long)member.colorValue);
    NSLog(@"%@", member.details);
    NSLog(@"%ld", (long)member.tags);
  }
  
  return cell;
}

@end

File: ClassicApp/TableViewController.swift
//
//  TableViewController.swift
//  ClassicApp
//
//  Created by Leo Dion on 2/8/22.
//

import UIKit

class TableViewController: UITableViewController {
  let repository : Repository = RepositoryImpl()
  var members : [Member]?
  var error : Error?
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    let nib = UINib(nibName: "TableViewCell", bundle: nil)
    self.tableView.register(nib, forCellReuseIdentifier: "reuseIdentifier")
    
    repository.getMembers { [weak self] result in
      switch result {
      case let .success(members):
        self?.members = members
      case let .failure(error):
        self?.error = error
      }
      self?.tableView.reloadData()
    }
  }
  
  // MARK: - Table view data source
  
  
  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    // #warning Incomplete implementation, return the number of rows
    return self.members?.count ?? 0
  }
  
  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let genericCell = tableView.dequeueReusableCell(withIdentifier: "reuseIdentifier", for: indexPath)
    
    guard let cell = genericCell as? TableViewCell else {
      return genericCell
    }
    
    if let member = self.members?[indexPath.row] {
      cell.nameLabel.text = member.firstName
      cell.nameLabel.textColor = member.uiColor
    }
    
    return cell
  }
  
}
